internal/crypto/cms/encoding/ber/ber.go

```diff
@@ -0,0 +1,301 @@
+		return nil, 0, nil, err
+	}
+
+	contentLen, r, err := decodeLength(r)
+	if err != nil {
+		return nil, 0, nil, err
+	}
+
+	return identifier, contentLen, r, nil
+}
+
+// decodeIdentifier decodes decodeIdentifier octets.
+//
+// Parameters:
+// r - The input byte slice from which the identifier octets are to be decoded.
+//
+// Returns:
+// []byte - The identifier octets decoded from the input byte slice.
+// []byte - The remaining part of the input byte slice after the identifier octets.
+// error - An error that can occur during the decoding process.
+//
+// Reference: ISO/IEC 8825-1: 8.1.2
+func decodeIdentifier(r []byte) ([]byte, []byte, error) {
+	if len(r) < 1 {
+		return nil, nil, asn1.SyntaxError{Msg: "decoding BER identifier octets: identifier octets is empty"}
+	}
+	offset := 0
+	b := r[offset]
+	offset++
+
+	// high-tag-number form
+	// Reference: ISO/IEC 8825-1: 8.1.2.4
+	if b&0x1f == 0x1f {
+		for offset < len(r) && r[offset]&0x80 == 0x80 {
+			offset++
+		}
+		if offset >= len(r) {
+			return nil, nil, asn1.SyntaxError{Msg: "decoding BER identifier octets: high-tag-number form with early EOF"}
+		}
+		offset++
+	}
+
+	if offset >= len(r) {
+		return nil, nil, asn1.SyntaxError{Msg: "decoding BER identifier octets: early EOF due to missing length and content octets"}
+	}
+	return r[:offset], r[offset:], nil
+}
+
+// decodeLength decodes length octets.
+// Indefinite length is not supported
+//
+// Parameters:
+// r - The input byte slice from which the length octets are to be decoded.
+//
+// Returns:
+// int - The length decoded from the input byte slice.
+// []byte - The remaining part of the input byte slice after the length octets.
+// error - An error that can occur during the decoding process.
+//
+// Reference: ISO/IEC 8825-1: 8.1.3
+func decodeLength(r []byte) (int, []byte, error) {
+	if len(r) < 1 {
+		return 0, nil, asn1.SyntaxError{Msg: "decoding BER length octets: length octets is empty"}
+	}
+	offset := 0
+	b := r[offset]
+	offset++
+
+	if b < 0x80 {
+		// short form
+		// Reference: ISO/IEC 8825-1: 8.1.3.4
+		contentLen := int(b)
+		subsequentOctets := r[offset:]
+		if contentLen > len(subsequentOctets) {
+			return 0, nil, asn1.SyntaxError{Msg: "decoding BER length octets: short form length octets value should be less or equal to the subsequent octets length"}
+		}
+		return contentLen, subsequentOctets, nil
+	} else if b == 0x80 {
+		// Indefinite-length method is not supported.
+		// Reference: ISO/IEC 8825-1: 8.1.3.6.1
+		return 0, nil, asn1.StructuralError{Msg: "indefinite length not supported"}
+	}
+
+	// long form
+	// Reference: ISO/IEC 8825-1: 8.1.3.5
+	n := int(b & 0x7f)
+	if n > 4 {
+		// length must fit the memory space of the int type (4 bytes).
+		return 0, nil, asn1.StructuralError{Msg: fmt.Sprintf("length of encoded data (%d bytes) cannot exceed 4 bytes", n)}
+	}
+	if offset+n >= len(r) {
+		return 0, nil, asn1.StructuralError{Msg: "decoding BER length octets: long form length octets with early EOF"}
```

Comment:
This should be an `asn1.SyntaxError`:
```suggestion
		return 0, nil, asn1.SyntaxError{Msg: "decoding BER length octets: long form length octets with early EOF"}
```